src/compiler/checker.ts

```diff
@@ -25970,12 +25970,17 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                             if (middleLength === 2) {
                                 if (elementFlags[startLength] & elementFlags[startLength + 1] & ElementFlags.Variadic) {
                                     // Middle of target is [...T, ...U] and source is tuple type
-                                    const targetInfo = getInferenceInfoForType(elementTypes[startLength]);
+                                    let targetInfo = getInferenceInfoForType(elementTypes[startLength]);
                                     if (targetInfo && targetInfo.impliedArity !== undefined) {
                                         // Infer slices from source based on implied arity of T.
                                         inferFromTypes(sliceTupleType(source, startLength, endLength + sourceArity - targetInfo.impliedArity), elementTypes[startLength]);
                                         inferFromTypes(sliceTupleType(source, startLength + targetInfo.impliedArity, endLength), elementTypes[startLength + 1]);
                                     }
+                                    else if ((targetInfo = getInferenceInfoForType(elementTypes[startLength + 1]))?.impliedArity !== undefined) {

```

Comment:
TODO: I might have to limit this to fixed-sized sources or something. This requires more test cases to exercise the improved behavior. It would help me slightly if I'd have a playground with this build - @jakebailey would you mind creating one? :)